Update Readme.md (#5)
DevExpress-Examples/uploadcontrol-an-example-of-use-t227032: Readme.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T227032)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
<!-- default badges end -->

> # Upload Control for ASP.NET MVC - How to upload multiple files at once

Reviewer: Why the two parts of the topic's header are capitalized according to different rules? I expect all headers to be capitalized in the same manner we do everywhere.

One more thing -- How to upload multiple files **at once**. "At once" is probably not the best choice here because it can be understood in more than one way:

1. At once -- right away, immediately. This is probably not what you originally meant
2. At once -- at a time, that is, in a single batch. This is, most likely the very thing you meant.
3. At once -- simultaneously. Yes, this is also probably the thing you've meant.

```suggestion
# Upload Control for ASP.NET MVC - How to upload multiple files simultaneously
```

OR

```suggestion
# Upload Control for ASP.NET MVC - How to upload multiple files at a fime
```

**Upd.**: We figured out personally that the second part of the readme title should not be capitalized as described in the following topic: http://tw.devexpress.devx/instructions/examples/formatting/how-to-format-examples/#title-in-readme

> This example demonstrates how to enable drag-and-drop support in the [Upload Control](https://docs.devexpress.com/AspNetMvc/8977/components/file-management/file-upload) extension and allow users to upload multiple files at once. After a user uploads files on the server, links to the uploaded files appear in the [Round Panel](https://docs.devexpress.com/AspNetMvc/8976/components/multi-use-site-extensions/roundpanel?p=netframework) below the Upload Control.

Reviewer: Likewise. Moreover, "once" here better serves as a beginning of the second sentence in this paragraph (Once == As soon as):

```suggestion
This example demonstrates how to enable drag-and-drop support in the [Upload Control](https://docs.devexpress.com/AspNetMvc/8977/components/file-management/file-upload) extension and allow users to upload multiple files simultaneously. Once a user uploads files on the server, links to the uploaded files appear in the [Round Panel](https://docs.devexpress.com/AspNetMvc/8976/components/multi-use-site-extensions/roundpanel?p=netframework) below the Upload Control.
```

![Upload Control](./images/uploadControl.gif "UploadControl")

## Overview:

Follow the steps below to allow users to upload multiple files and display links to uploaded files on a web page:

1. Add the [Upload Control extension](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlExtension?p=netframework) to a view. To do this, you can use the [Insert Extension Wizard](https://docs.devexpress.com/AspNetMvc/14551/whats-installed/insert-devexpress-mvc-extension-wizard) or write the extension code manually.

2. In the *Global.asax* or *RouteConfig* file, add the **.ashx** file extension to route exceptions:

    ```cs
    routes.IgnoreRoute("{resource}.axd/{*pathInfo}");
    routes.IgnoreRoute("{resource}.ashx/{*pathInfo}");
    ```

3. Place the [Round Panel extension](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.RoundPanelExtension?p=netframework) below the Upload Control:

    ```razor
    @Html.DevExpress().RoundPanel(settings => {
        settings.Name = "RoundPanel";
        settings.HeaderText = "Uploaded Files";
        settings.SetContent(() => { ViewContext.Writer.Write("<div id='fileContainer'/>"); });
    }).GetHtml()
    ```

> 4. Configure settings of the Upload Control extension. Set the extension's [UploadMode](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlSettings.UploadMode?p=netframework) property to `true` to enable advanced upload mode. To allow users to upload multiple files, set the [EnableMultiSelect](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableMultiSelect?p=netframework) property to `true`. The [EnableDragAndDrop](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableDragAndDrop) property specifies whether to enable drag-and-drop support.

Reviewer 2: ```suggestion
4. Configure settings of the Upload Control extension. Set the extension's [UploadMode](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlSettings.UploadMode?p=netframework) property to `[Advanced](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadControlUploadMode)` to enable advanced upload mode. To allow users to upload multiple files, set the [EnableMultiSelect](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableMultiSelect?p=netframework) property to `true`. The [EnableDragAndDrop](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableDragAndDrop) property specifies whether to enable drag-and-drop support.
```

Author: ```suggestion
4. Configure settings of the Upload Control extension. Set the extension's [UploadMode](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlSettings.UploadMode?p=netframework) property to [Advanced](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadControlUploadMode) to enable advanced upload mode. To allow users to upload multiple files, set the [EnableMultiSelect](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableMultiSelect?p=netframework) property to `true`. The [EnableDragAndDrop](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableDragAndDrop) property specifies whether to enable drag-and-drop support.
```

> 4. Configure settings of the Upload Control extension. Set the extension's [UploadMode](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlSettings.UploadMode?p=netframework) property to [Advanced](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadControlUploadMode) to enable advanced upload mode. To allow users to upload multiple files, set the [EnableMultiSelect](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableMultiSelect?p=netframework) property to `true`. The [EnableDragAndDrop](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableDragAndDrop) property specifies whether to enable drag-and-drop support.

Reviewer: I've always been a consistent adversary of the pattern "whether to do something". Is there any specific reason for using it?

What does this property specify? It specifies whether support for drag-and-drop operation is active. Nothing more. It just declares if the said functionality is disabled or enabled, it does not allow your component to decide on its own.

We should be extremely to-the-point in everything we write. Phrases like "whether to enable" leave too much room for interpretation, and use more words as well.

Whether == if, if is shorter. However, I don't have any objections against the word "whether" itself, just this pattern "whether to do something".

Can we simplify this sentence further? We can. Can we deduce "enabled drag-and-drop operations" from "enabled drag-and-drop support"? Yes, we can. While the final result may not be much shorter in terms of the overall character count, it has more ties with operations made by users directly.

```suggestion
4. Configure settings of the Upload Control extension. Set the extension's [UploadMode](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlSettings.UploadMode?p=netframework) property to [Advanced](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadControlUploadMode) to enable advanced upload mode. To allow users to upload multiple files, set the [EnableMultiSelect](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableMultiSelect?p=netframework) property to `true`. The [EnableDragAndDrop](https://docs.devexpress.com/AspNet/DevExpress.Web.UploadAdvancedModeSettings.EnableDragAndDrop) property specifies if drag-and-drop operations are enabled.
```

    ```razor
    @using (Html.BeginForm("UploadControlUpload", "Home")) {
        @Html.DevExpress().UploadControl(settings => {
            // ...
        }).GetHtml()
    }
    ```

6. In the action method, call the Upload Control's [GetUploadedFiles](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlExtension.-l------------y-) method overload to get the uploaded files and handle the server-side [FilesUploadComplete](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxUploadControl.FilesUploadComplete) event:

    ```cs
    public ActionResult UploadControlUpload() {
        string[] errors;
        UploadedFile[] files = UploadControlExtension.GetUploadedFiles("UploadControl",
            MyUploadControlValidationSettings.Settings, out errors, (s, e) => { },
            UploadControl_FilesUploadComplete);
        return null;
    }
    ```

> 7. In the `FilesUploadComplete` event handler, save the uploaded files on the server. Write names of uploaded files and their URLs in a string. Assign this string to the [CallbackData](https://docs.devexpress.com/AspNet/DevExpress.Web.FilesUploadCompleteEventArgs.CallbackData) event argument to pass the string to the client:

Reviewer: > Write names of uploaded files and their URLs **in** a string.

My first question is addressed to the used preposition. "In" normally answers the aspect of place. Where? In a string. "To", respectively, conveys the aspect of target. Where? To a string.

This case urges us to use a combined preposition "into". But, since we decided not to use it, we have to choose the most important aspect. And, in this particular case, the most important aspect is the aspect of target. Because the string is just a container for your information.

I understand that your code example below describes how to do it, yet still would have probably considered adding some information on "how to write names of uploaded files and their URLs"? In what format? What should follow what? What separator characters should one insert between entries within a string? Etc., etc.

I do not insist, but at least consider this...

```suggestion
7. In the `FilesUploadComplete` event handler, save the uploaded files on the server. Write names of uploaded files and their URLs to a string. Assign this string to the [CallbackData](https://docs.devexpress.com/AspNet/DevExpress.Web.FilesUploadCompleteEventArgs.CallbackData) event argument to pass the string to the client:
```

Author: I do not think we should tell our users how exactly they should pass information to the client to implement this particular use case. 

As I see, the main goal of the example is to illustrate the basic concepts of working with the Upload Control. And it doesn't really matter what we send from the server to the client in this example, it's important that it can be done. If the reader is really interested in this particular use case, he or she can review the code example below.

Anyway, thank you for your comment. I appreciate your desire to improve our learning materials.

Reviewer: Alright, I understand your point. Focus is important, purpose is important. Thank you for your critical attitude. It's invaluable.

            var file = data[i];
            var url = data[i + 1];
            var link = document.createElement('A');
            link.innerHTML = file;
            link.setAttribute('href', url);
            link.setAttribute('target', '_blank');
            var fileContainer = document.getElementById('fileContainer');
            fileContainer.appendChild(link);
            fileContainer.appendChild(document.createElement('BR'));
        }
    }
    ```

## Files to Review

* [HomeController.cs](./CS/UploadControlExample/Controllers/HomeController.cs) (VB: [HomeController.vb](./VB/UploadControlExample/Controllers/HomeController.vb))
* [UploadControlValidationSettings.cs](./CS/UploadControlExample/Models/UploadControlValidationSettings.cs) (VB: [UploadControlValidationSettings.vb](./VB/UploadControlExample/Models/UploadControlValidationSettings.vb))
* [Index.cshtml](./CS/UploadControlExample/Views/Home/Index.cshtml) (VB: [Index.vbhtml](./VB/UploadControlExample/Views/Home/Index.vbhtml))

## Documentation

* [Upload Modes](https://docs.devexpress.com/AspNet/9886/components/file-management/file-upload/concepts/upload-modes)

## More Examples

> * [Upload Control for ASP.NET MVC - How to bind uploaded files to a model](https://github.com/DevExpress-Examples/uploadcontrol-with-model-binding-support-registration-form-t185980)

Reviewer: These links have the same heading capitalization issues as the current topic. Is it intentional?

Reviewer: Yes, we figured out that this is intentional. The question is revoked.
